File: bindings/flutter/lib/src/catalog.dart
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import 'dart:convert';
import 'dart:ffi';

import 'package:ffi/ffi.dart';
import 'package:meta/meta.dart';

import 'bindings/bindings.dart' as raw;
import 'bindings/native_library.dart';
import 'error_capture.dart';
import 'job_runner.dart';
import 'model.dart';
import 'models/model_info.dart';
import 'native_strings.dart';

/// Filter accepted by [Catalog.listModels].
class CatalogFilter {
  const CatalogFilter({
    this.task,
    this.cachedOnly = false,
    this.loadedOnly = false,
    this.maxSizeBytes,
    this.compatibleOnly = true,
  });

  final String? task;
  final bool cachedOnly;
  final bool loadedOnly;
  final int? maxSizeBytes;
  final bool compatibleOnly;

  Map<String, Object?> toJson() {
    return <String, Object?>{
      if (task != null) 'task': task,
      if (cachedOnly) 'cached_only': cachedOnly,
      if (loadedOnly) 'loaded_only': loadedOnly,
      if (maxSizeBytes != null) 'max_size_bytes': maxSizeBytes,
      'compatible_only': compatibleOnly,
    };
  }
}

/// The model catalog. Handle is borrowed from the manager and remains valid
/// for the manager's lifetime; there is no explicit release.
class Catalog {
  /// Internal constructor — call [FoundryLocal.catalog] to get an instance.
  @internal
  Catalog.internal(this._handle);

  final int _handle;

  /// List models. When [filter] is null every catalog entry is returned; a
  /// filter restricts the result to models the device can run.
  Future<List<ModelInfo>> listModels({CatalogFilter? filter}) async {
    final bindings = NativeLibrary.instance.bindings;
    final filterJson = filter == null ? null : jsonEncode(filter.toJson());

    final result = await withNullableCString<Future<Map<String, Object?>>>(
      filterJson,
      (filterPtr) => runSimpleJob(
        (completionPtr, userData, outJob) => bindings
            .flm_catalog_list_models_async(
                _handle, filterPtr, completionPtr, userData, outJob),
      ),
    );
    final list = (result['models'] as List?)?.cast<Map<String, Object?>>() ??
        const <Map<String, Object?>>[];
    return list.map(ModelInfo.fromJson).toList(growable: false);
  }

  /// Resolve a model by alias (`qwen2.5-0.5b`). This may hit the network to
  /// resolve the alias if the catalog has not been fetched yet.
  Future<Model> getModel(String alias) async {
    final bindings = NativeLibrary.instance.bindings;
    final result = await withCString<Future<Map<String, Object?>>>(
      alias,
      (aliasPtr) => runSimpleJob(
        (completionPtr, userData, outJob) => bindings
            .flm_catalog_get_model_async(
                _handle, aliasPtr, completionPtr, userData, outJob),
      ),
    );
    final handle = (result['model_handle'] as num?)?.toInt() ?? 0;
    if (handle == raw.FLM_INVALID_HANDLE) {
      throw StateError(
          'Catalog returned an invalid model handle for alias "$alias".');
    }
    return Model.fromHandle(handle);
  }

  /// Resolve a specific variant by its full id, bypassing automatic selection.
  Future<Model> getModelById(String modelId) async {
    final bindings = NativeLibrary.instance.bindings;
    final result = await withCString<Future<Map<String, Object?>>>(
      modelId,
      (idPtr) => runSimpleJob(
        (completionPtr, userData, outJob) => bindings
            .flm_catalog_get_model_by_id_async(
                _handle, idPtr, completionPtr, userData, outJob),
      ),
    );
    final handle = (result['model_handle'] as num?)?.toInt() ?? 0;
    if (handle == raw.FLM_INVALID_HANDLE) {
      throw StateError(
          'Catalog returned an invalid model handle for id "$modelId".');
    }
    return Model.fromHandle(handle);
  }

  /// Models already present in the local cache. Serves from disk, so it is
  /// safe to call during startup before any network is available.
  List<ModelInfo> listCachedModels() {
    final bindings = NativeLibrary.instance.bindings;
    final out = calloc<Pointer<Char>>();
    try {
      final status =
          bindings.flm_catalog_list_cached_models_json(_handle, out);
      checkStatus(status,
          fallbackMessage: 'flm_catalog_list_cached_models_json failed');
      try {
        final s = takeOutString(out.value);
        if (s.isEmpty) return const <ModelInfo>[];
        final decoded = jsonDecode(s);
        final list =
            (decoded is List ? decoded : (decoded as Map)['models'] as List?);
        return (list ?? const <Object?>[])
            .cast<Map<String, Object?>>()
            .map(ModelInfo.fromJson)
            .toList(growable: false);
      } finally {
        bindings.flm_string_free(out.value);
      }
    } finally {
      calloc.free(out);
    }
  }

  /// Bytes currently consumed by the model cache.
  int get cacheSizeBytes {
    final bindings = NativeLibrary.instance.bindings;
    final out = calloc<Int64>();
    try {
      final status =
          bindings.flm_catalog_get_cache_size_bytes(_handle, out);
      checkStatus(status,
          fallbackMessage: 'flm_catalog_get_cache_size_bytes failed');
      return out.value;
    } finally {
      calloc.free(out);
    }
  }
}

File: bindings/ios/Sources/FoundryLocal/Catalog.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import Foundation
import FoundryLocalMobile

/// Read-only view of the model catalog. Obtained from ``FoundryLocal/catalog``.
///
/// Catalog handles are borrowed from the owning ``FoundryLocal``: releasing them is a
/// no-op, and they become invalid when the manager is released.
public final class Catalog: @unchecked Sendable {
    public let handle: flm_catalog

    init(handle: flm_catalog) {
        self.handle = handle
    }

    /// List catalog models, optionally filtered.
    public func listModels(_ filter: CatalogFilter = .init()) async throws -> [ModelInfo] {
        let filterJSON = filter.encodeAsJSON()
        return try await runAsyncJob(
            decode: { job in
                let json = try takeJobResultJSON(job)
                return try flmJSONDecoder.decode(CatalogListResult.self, from: Data(json.utf8)).models
            },
            submit: { [handle] userData, _, onComplete, outJob in
                filterJSON.withCString { filterPtr in
                    flm_catalog_list_models_async(handle, filterPtr, onComplete, userData, outJob)
                }
            }
        )
    }

    /// Resolve a model by its catalog alias (e.g. `"qwen2.5-0.5b"`). For a package
    /// this returns the package handle, not any specific variant.
    public func model(alias: String) async throws -> Model {
        try await runAsyncJob(
            decode: { job in
                let json = try takeJobResultJSON(job)
                let parsed = try flmJSONDecoder.decode(CatalogGetResult.self, from: Data(json.utf8))
                return Model(handle: flm_model(parsed.modelHandle))
            },
            submit: { [handle] userData, _, onComplete, outJob in
                alias.withCString { aliasPtr in
                    flm_catalog_get_model_async(handle, aliasPtr, onComplete, userData, outJob)
                }
            }
        )
    }

    /// Resolve a model by its unique id, bypassing automatic variant selection. Useful
    /// when the app is pinning a specific variant across sessions.
    public func model(id: String) async throws -> Model {
        try await runAsyncJob(
            decode: { job in
                let json = try takeJobResultJSON(job)
                let parsed = try flmJSONDecoder.decode(CatalogGetResult.self, from: Data(json.utf8))
                return Model(handle: flm_model(parsed.modelHandle))
            },
            submit: { [handle] userData, _, onComplete, outJob in
                id.withCString { idPtr in
                    flm_catalog_get_model_by_id_async(handle, idPtr, onComplete, userData, outJob)
                }
            }
        )
    }

    /// Models already present in the local cache. Serves from disk, so it is
    /// synchronous and safe to call before any network is available.
    public func cachedModels() throws -> [ModelInfo] {
        let json = try readJSON { flm_catalog_list_cached_models_json(handle, $0) }
        return try flmJSONDecoder.decode([ModelInfo].self, from: Data(json.utf8))
    }

    /// Bytes currently used by the model cache.
    public var cacheSizeBytes: Int64 {
        var bytes: Int64 = 0
        return flm_catalog_get_cache_size_bytes(handle, &bytes) == FLM_OK ? bytes : 0
    }
}

/// Filter for ``Catalog/listModels``.
public struct CatalogFilter: Sendable {
    public var task: String?
    public var cachedOnly: Bool = false
    public var loadedOnly: Bool = false
    public var maxSizeBytes: Int64?
    public var compatibleOnly: Bool = false

    public init(
        task: String? = nil,
        cachedOnly: Bool = false,
        loadedOnly: Bool = false,
        maxSizeBytes: Int64? = nil,
        compatibleOnly: Bool = false
    ) {
        self.task = task
        self.cachedOnly = cachedOnly
        self.loadedOnly = loadedOnly
        self.maxSizeBytes = maxSizeBytes
        self.compatibleOnly = compatibleOnly
    }

    func encodeAsJSON() -> String {
        var payload: [String: Any] = [:]
        if let task { payload["task"] = task }
        if cachedOnly { payload["cached_only"] = true }
        if loadedOnly { payload["loaded_only"] = true }
        if let maxSizeBytes { payload["max_size_bytes"] = maxSizeBytes }
        if compatibleOnly { payload["compatible_only"] = true }
        return payload.jsonString() ?? "{}"
    }
}
